feat: add Prometheus metrics integration for KvStats

keivenchang opened the pull request:
#### Overview:

Expose KvStats metrics to Prometheus format for monitoring KV cache utilization and performance.

#### Details:

- Added `KvStatsPrometheusGauges` struct to manage four key metrics: active blocks, total blocks, GPU cache usage percentage, and prefix cache hit rate
- Integrated Prometheus gauge updates into `WorkerMetricsPublisher::publish()` using RwLock for optimal read performance in the hot path
- Registered metrics with Component's MetricsRegistry with standardized "kvstats_" prefix naming convention
- Added comprehensive integration test to verify metrics registration and value updates
- Moved test utilities to common.rs for better code organization
- Example of the new Prometheus metrics:
```
ubuntu@keivenc-linux:~/dynamo$ curl -s localhost:8081/metrics | grep -v '^#' | grep -v bucket | sort | grep kvstats
dynamo_component_kvstats_active_blocks{dynamo_component="backend",dynamo_namespace="dynamo"} 0
dynamo_component_kvstats_gpu_cache_usage_percent{dynamo_component="backend",dynamo_namespace="dynamo"} 0.00007755545084364712
dynamo_component_kvstats_gpu_prefix_cache_hit_rate{dynamo_component="backend",dynamo_namespace="dynamo"} 0
dynamo_component_kvstats_total_blocks{dynamo_component="backend",dynamo_namespace="dynamo"} 12894
```

#### Where should the reviewer start?

- `lib/llm/src/kv_router/publisher.rs` - Core implementation of KvStatsPrometheusGauges and integration with WorkerMetricsPublisher
- `lib/runtime/src/metrics/prometheus_names.rs` - Standardized metric name definitions

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

DIS-506

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Exposes Prometheus metrics for KV stats, including active blocks, total blocks, GPU cache usage %, and GPU prefix cache hit rate.
  - Endpoint creation now supports optional metric labels for improved observability.
  - Metrics are registered before endpoint startup to ensure immediate availability.

- Tests
  - Added integration tests validating KV stats gauge updates and Prometheus output formatting.

- Chores
  - Introduced standardized metric names and grouping for KV stats to streamline monitoring setup.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
feat: add Prometheus metrics integration for KvStats

- Add KvStatsPrometheusGauges struct to manage Prometheus gauges for KV cache metrics
- Integrate gauge updates into WorkerMetricsPublisher.publish() method
- Register metrics with Component's MetricsRegistry during initialization
- Add kvstats metric name constants in prometheus_names.rs
- Use RwLock for performance optimization in hot path
- Add integration test for KvStats Prometheus metrics
- Move test utilities to common.rs for reusability

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
refactor: make register_prometheus_metrics idempotent

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

nnshah1 reviewed (approved):
LGTM!

keivenchang committed:
add integration feature so it does not run on CI

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
refactor: replace println with tracing::debug for metrics output

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
ignore test_kvstats_prometheus_gauge_updates

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
minor re-org of test groups

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on lib/llm/src/common.rs:
```diff
@@ -15,3 +15,18 @@
+
+    /// Creates a test DistributedRuntime for integration testing.
+    /// Uses from_current to leverage the existing tokio runtime in tests.
+    pub async fn create_test_drt_async() -> DistributedRuntime {
```
You are absolutely right, the one in distributed.rs was merged in 2 days ago (before it was available when I started this PR, before the merge). Now that it's available on main, I'll use it. Thanks for the reminder!

keivenchang committed:
consolidate create_test_drt_async

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
reorganize the tests

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix the create_test_drt_async feature gating

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix the create_test_drt_async imports

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
add ignore to tests

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
Merge branch 'main' into keivenchang/DIS-506__expose-kvstats-to-prometheus-format-metrics

Signed-off-by: Keiven C <213854356+keivenchang@users.noreply.github.com>

keivenchang committed:
fix: resolve merge conflict in prometheus_names.rs

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

rmccorm4 commented on lib/bindings/python/rust/llm/kv.rs:
@keivenchang can you please add a snippet output of the net new metrics to PR description: `curl host:port/metrics | grep <new_kv_metrics>`? Helps give some immediate context/understanding on the addition.

PeaBrane commented on lib/llm/src/kv_router/publisher.rs:
```diff
@@ -482,22 +483,118 @@ enum RawKvEvent {
+    /// Prometheus gauges for KvStats metrics
+    /// We use Option<Arc<KvStatsPrometheusGauges>> for lock-free reads after initialization
+    /// The Arc is set once during register_prometheus_metrics and then only read
+    prometheus_gauges: std::sync::RwLock<Option<Arc<KvStatsPrometheusGauges>>>,
```
Can't you make this an `OnceLock<KvStatsPrometheusGauges>`. I'm not sure if `Arc` or `RwLock` is needed as you seem to be only using this on the main thread.

keivenchang replied:
Done. You're right, this simplifies a lot of the logic. And Arc/RwLock is not needed. Prometheus Gauges are also thread-safe as well.

PeaBrane commented on lib/llm/src/kv_router/publisher.rs:
```diff
@@ -482,22 +483,118 @@ enum RawKvEvent {
+
+        // Update Prometheus gauges using read lock for better performance
+        // This is the hot path - we only read the Arc, no writes needed
+        #[allow(clippy::collapsible_if)] // Necessary to keep guard alive
+        if let Ok(guard) = self.prometheus_gauges.read() {
+            if let Some(gauges) = guard.as_ref() {
+                gauges.update_from_kvstats(&metrics.kv_stats);
+            }
+        }
```
stale suggestion, see below comment

keivenchang replied:
This code is a lot more simplified after using OnceLock. Thanks for the suggestion.

keivenchang committed:
refactor: use OnceLock for prometheus_gauges instead of RwLock

RwLock is not necessary since gauges are initialized once and then only read.
OnceLock simplifies the code and improves performance on the hot path.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

kthui reviewed (approved):
If I understand correctly, this feature adds the `KvStatsPrometheusGauges` struct that tracks 4 metrics supplied by the `KvStats` that is already a part of the `ForwardPassMetrics`.

One question: I wonder if the `Arc` around the `KvStatsPrometheusGauges` is needed, unless it lives longer than the `WorkerMetricsPublisher`?

keivenchang committed:
Merge branch 'main' into keivenchang/DIS-506__expose-kvstats-to-prometheus-format-metrics

Signed-off-by: Keiven C <213854356+keivenchang@users.noreply.github.com>

keivenchang committed:
refactor: simplify KvStatsPrometheusGauges

Removed unnecessary Arc and Option wrappers since gauges are never shared
independently and are always initialized.

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang reviewed (commented):
Hi @kthui I looked at this more closely and yes, it's only accessed here so it does not need an Arc. I was confused because the Sender and Receiver were using the Arc.

In addition, I made KvStatsPrometheusGauges not to use Option (made it mandatory). That also simplified the code quite a bit.
